orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/PublishDeliveryConfigTask.kt

```diff
@@ -0,0 +1,157 @@
+        throw IllegalArgumentException("scmType, project and repository are required fields in the stage if there's no git trigger.")
+      }
+    }
+
+    val manifestLocation = "${context.scmType}://${context.project}/${context.repository}/.netflix/${context.directory
+      ?: ""}/${context.manifest}@${context.ref}"
+
+    return try {
+      log.info("Retrieving keel manifest at $manifestLocation")
+      val deliveryConfig = scmService.getDeliveryConfigManifest(
+        context.scmType, context.project, context.repository, context.directory, context.manifest, context.ref)
+
+      log.info("Publishing manifest ${context.manifest} to keel on behalf of $user")
+      keelService.publishDeliveryConfig(deliveryConfig, user)
+
+      TaskResult.builder(ExecutionStatus.SUCCEEDED).context(emptyMap<String, Any?>()).build()
+    } catch (e: RetrofitError) {
+      when {
+        e.kind == RetrofitError.Kind.NETWORK -> {
+          // retry if unable to connect
+          log.error("network error talking to downstream service, attempt ${context.attempt} of ${context.maxRetries}: ${e.friendlyMessage}")
+          buildRetry(context)
+        }
+        e.response?.status == HttpStatus.NOT_FOUND.value() -> {
+          // just give up on 404
+          val errorDetails = "404 response from downstream service, giving up: ${e.friendlyMessage}"
+          log.error(errorDetails)
+          buildError(errorDetails)
+        }
+        else -> {
+          // retry on other status codes
+          log.error("HTTP error talking to downstream service, attempt ${context.attempt} of ${context.maxRetries}: ${e.friendlyMessage}")
+          buildRetry(context)
+        }
+      }
+    } catch (e: Exception) {
+      log.error("Unexpected exception while executing {}, aborting.", javaClass.simpleName, e)
+      buildError(e.message)
+    }
+  }
+
+  override fun getBackoffPeriod() = TimeUnit.SECONDS.toMillis(30)
+
+  override fun getTimeout() = TimeUnit.SECONDS.toMillis(180)
+
+  private fun buildRetry(context: PublishDeliveryConfigContext) =
+    TaskResult.builder(ExecutionStatus.RUNNING).context(context.incrementAttempt().toMap()).build()
+
+  private fun buildError(errorDetails: String?) =
+    TaskResult.builder(ExecutionStatus.TERMINAL).context(mapOf("error" to errorDetails)).build()
+
+  val RetrofitError.friendlyMessage: String
+    get() = "HTTP ${response.status} ${response.url}: ${cause?.message ?: message}"
+
+  data class PublishDeliveryConfigContext(
+    var scmType: String? = null,
+    var project: String? = null,
+    var repository: String? = null,
+    var directory: String? = null, // as in, the directory *under* whatever manifest base path is configured in igor (e.g. ".netflix")
+    var manifest: String? = "spinnaker.yml",
+    var ref: String? = null,
+    var attempt: Int = 1,
+    val maxRetries: Int? = MAX_RETRIES
+//    val forceRepublish: Boolean? = false // TODO
```

Comment:
I think we probably don't need this, it's not a big deal to always republish ( i have found in practice i always want to republish)